### ISSUE TITLE
threads, ssl init

### ISSUE THREAD

**Issue by imagineosc** [warmcat/libwebsockets#945]
If one thread sequentially initiates and connects N web sockets (with their own contexts), which are each immediately serviced in their own thread, then I experience segfaults traced to lws_context_init_ssl_library.  I think it has something to do with new connections being initiated in one thread while a different thread is servicing a connection that is in progress.

I have also tried creating, initializing, and servicing each connection in its own dedicated thread.  That still crashes.

The only configuration that seems to avoid the crashes is to begin all of the connections in a single thread, and then start the servicing threads.

I have not been able to find clear documentation on what the thread safety guarantees are for libwebsockets, or what the recommended idioms are for creating and servicing in multiple threads. The most i have found are comments to the effect of "one context per process," but I have tried that (or rather, one context per thread), and still experienced crashing.  And if you really do mean "one context per process," instead of one per thread, then the system is way too fragile.

I realize that openssl is terrible when it comes to thread safety. But for this to be a robust solution, it needs to work around the openssl limitations (or switch to a better library).

**Comment by lws-team** [comment]
It's probably better to hold on the judgement and advice how to solve (for both lws and openssl) until you have less doubts about what is actually going on.

 - older openssl has a poblem that its library init and descruction is processwide.  That creates a situation that multiple instaces of the library in a single process conflict.  There's nothing to be done about that.

 - due to the many complaints they fixed this by adding new apis at some point.  If lws sees that these are available, it will use them.  Then openssl should be able to coexist IF the other library including it also uses the new init and destruction apis.

 - lws is SINGLETHREADED.  There is NO LOCKING.  This reflects the nonblocking, serialized service approach.  Many things use this approach, eg, libuv itself.  I get it you are used to throwing a thread on every connection and blocking it.  That is not the approach used here, and your app design needs to reflect that.  There are other ws libs out there that also use threaded / blocking approach, if you feel that's the only way you can do what you are doing, please use some other library.

 - One of the jobs of the lws context is to perform fast mapping of fds to wsi for the whole fd space of your process.  Multiple processes will not conflict, since in a process an fd can only be in use once, but you would have to a) provide service for them individually, b) take the bloat that it allocated fd tables, or restrict how many fds it deals with using the context init info struct, c) take other once-per-context bloat each time, and d) risk openssl conflicting with itself outside of lws.

In short you're just applying your own app design to the situation and ignoring what lws is telling you, and blaming lws for openssl problems.  You can get out of that by adapting your design or use a different library.

Also, from an issue reporting POV, no point telling me there is a segfault somewhere.  Why not give me a backtrace so I can see the exact line?  Perhaps your actual problem is not even related to your assumptions.

**Comment by imagineosc** [comment]
I'm very familiar with nonblocking, event-driven io. I have over a decade of deep experience with that paradigm, having  built web crawlers, distributed file systems, rpc, etc, all around epoll.

But sometimes it's not appropriate. Your code is super fragile and poorly documented and your response to suggestions for improvements and clarifications is to make ad hominem attacks.

Before you wrote the above response, I was in the process of trying to refactor to a shared context model.  And then I realized that the context needs the protocols specified at creation time. That's just not architecturally reasonable if it's global.

**Comment by lws-team** [comment]
> Your code is super fragile

 Like the segfault, no point waving your hands and "making ad hominem attacks".  Where is lws fragile?  Because it uses openssl and older openssl supports only one instance of itself per process?  Libcurl is "fragile" the same way?

> And then I realized that the context needs the protocols specified at creation time ... That's just not architecturally reasonable if it's global.

It's only global to the process if you have older openssl with its one-per-process init problem.  Multiple contexts per se will not conflict with each other in the same way that openssl's old init / destruction does.  You got "a segfault" and extrapolated from that, but to understand what actually happened you should raise an issue with a backtrace.  You did not do so, just dumped a bunch of confused complaints and now attacks.  You can either back up, provide data and wait until you can show some complaint is justified (on a FOSS project though, is it ever **useful**), use a different library with threaded model, or continue wasting your and my time with posts the same as the two above.
